ud_dataset/ud_dataset.py

```diff
@@ -0,0 +1,132 @@
+
+    class Dataset:
+        def __init__(self, data_file: TextIO, max_tokens: int | None = None) -> None:
+            # Load the data
+            self._size = 0
+            self.forms = []
+            self.poss = []
+            self.genders = []
+
+            for sentence in conllu.parse_incr(data_file):
+                for token in sentence:
+                    form = token.get("form")
+                    pos = token.get("upostag")
+                    gender = None
+                    if 'feats' in token and token['feats'] is not None:
+                        feats = token['feats']
+                        if 'Gender' in feats:
+                            gender = feats['Gender']
+
+                            if gender == "Masc":
+                                gender = Gender.MASCULINE
+                            elif gender == "Fem":
+                                gender = Gender.FEMININE
+                            else:
+                                gender = Gender.OTHER
+                    self.forms.append(form)
+                    self.poss.append(pos)
+                    self.genders.append(gender)
+                    self._size += 1
+                    if max_tokens is not None and self._size >= max_tokens:
+                        break
+
+        def __len__(self) -> int:
+            return self._size
+
+        def __getitem__(self, index: int) -> str:
+            return self.forms[index]
+
+        @property
+        def text(self) -> list[str]:
+            """
+            Get the unannotated text of the dataset, as a list of tokens.
+            :return: The unannotated text.
+            """
+            return self.forms
+
+        def get_unique_nouns(self) -> set[str]:
+            """
+            Extracts the set of unique nouns from the dataset.
+            :return: The set of unique nouns.
+            """
+            nouns = [form for (form, pos) in zip(self.forms, self.poss) if pos == "NOUN"]
+
+            # Filter for errors: consider a word to be a noun only if it appears more than 4 times annotated as a NOUN.
+            counts = Counter(nouns)
+            relevant_nouns = [noun for noun in counts if counts[noun] > 4]
+            noun_set = set(relevant_nouns)
+            return noun_set
+
+    def __init__(self, max_tokens=None) -> None:
+        for dataset_name, dataset in zip(["train-lt", "dev", "test"], ["train", "dev", "test"]):
+            # for dataset_name, dataset in zip(["train-ca", "dev", "test"], ["train", "dev", "test"]):
+            filename = f"cs_pdt-ud-{dataset_name}.conllu"
+            path = DATA_DIR / filename
+            if not os.path.exists(path):
+                print("Downloading dataset {}...".format(filename), file=sys.stderr)
+                urllib.request.urlretrieve(
+                    "{}/{}".format(self._URL, filename), filename="{}.tmp".format(path))
```

Comment:
I think that f-strings are more readable (except some rare cases, but this is not one of them).
```suggestion
                print(f"Downloading dataset {filename}...", file=sys.stderr)
                urllib.request.urlretrieve(f"{self._URL}/{filename}", filename=f"{path}.tmp")
```